pkg/controllers/nodeoverlay/controller.go

```diff
@@ -0,0 +1,268 @@
+	}
+	if err := c.kubeClient.List(ctx, nodePoolList); err != nil {
+		return reconcile.Result{}, fmt.Errorf("listing nodepools, %w", err)
+	}
+
+	overlayWithRuntimeValidationFailure := map[string]error{}
+	overlaysWithConflict := []string{}
+	temporaryStore := NewInstanceTypeStore()
+	nodePoolToInstanceTypes := map[string][]*cloudprovider.InstanceType{}
+
+	for _, np := range nodePoolList.Items {
+		its, err := c.cloudProvider.GetInstanceTypes(ctx, &np)
+		if err != nil {
+			return reconcile.Result{}, fmt.Errorf("listing instance types, %w", err)
+		}
+		nodePoolToInstanceTypes[np.Name] = its
+	}
+
+	overlayList.OrderByWeight()
+	for x := range overlayList.Items {
+		if err := overlayList.Items[x].RuntimeValidate(ctx); err != nil {
+			overlayWithRuntimeValidationFailure[overlayList.Items[x].Name] = err
+			continue
+		}
+
+		// Due to reserved capacity type offering being dynamically injected as part of the GetInstanceTypes call
+		// We will need to make sure we are validating against each nodepool to make sure. This will ensure that
+		// overlays that are targeting reserved instance offerings will be able to apply the offering.
+		for y := range nodePoolList.Items {
+			if !c.filterAndStoreValidInstanceTypesOverride(temporaryStore, nodePoolList.Items[y], nodePoolToInstanceTypes[nodePoolList.Items[y].Name], overlayList.Items[x]) {
+				overlaysWithConflict = append(overlaysWithConflict, overlayList.Items[x].Name)
+			}
+		}
+	}
+	temporaryStore.verifiedNodePools = lo.Map(nodePoolList.Items, func(np v1.NodePool, _ int) string {
+		return np.Name
+	})
+
+	err, requeue := c.updateOverlayStatuses(ctx, overlayList.Items, overlaysWithConflict, overlayWithRuntimeValidationFailure)
+	if requeue {
+		return reconcile.Result{Requeue: true}, nil
+	}
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("updating nodeoverlay statuses, %w", err)
+	}
+
+	c.instanceTypeStore.UpdateStore(temporaryStore)
+	return reconcile.Result{RequeueAfter: 6 * time.Hour}, nil
+}
+
+func (c *Controller) Register(_ context.Context, m manager.Manager) error {
+	b := controllerruntime.NewControllerManagedBy(m).
+		Named(c.Name()).
+		// The reconciled overlay does not matter in this case as one reconcile loop
+		// will compare every overlay against every other overlay in the cluster.
+		For(&v1alpha1.NodeOverlay{}).
+		Watches(&v1.NodePool{}, NodeOverlayEventHandler(c.kubeClient)).
+		WithEventFilter(predicate.GenerationChangedPredicate{}).
+		WithOptions(controller.Options{
+			MaxConcurrentReconciles: 1,
+			RateLimiter:             reasonable.RateLimiter(),
+		})
+
+	for _, nodeClass := range c.cloudProvider.GetSupportedNodeClasses() {
+		b.Watches(nodeClass, NodeOverlayEventHandler(c.kubeClient))
+	}
+
+	return b.Complete(c)
+}
+
+func (c *Controller) filterAndStoreValidInstanceTypesOverride(store *InstanceTypeStore, nodePool v1.NodePool, its []*cloudprovider.InstanceType, overlay v1alpha1.NodeOverlay) bool {
```

Comment:
We are now going to validate every instance type and then apply the updates to the instance type store